frame/collective/src/lib.rs

```diff
@@ -279,55 +531,96 @@ decl_module! {
+		///   - approved:          65.95 + M * 0.226 + P2 * 0.487 + B * 0.005 µs (min squares analysis)
+		/// # </weight>
+		#[weight = (
+			weight_for::close(
+				T::DbWeight::get(),
+				T::MaxMembers::get(), // `M`
+				*proposal_weight_bound, // `P1`
+				T::MaxProposals::get() // `P2`
+			),
+			DispatchClass::Operational
+		)]
+		fn close(
+			origin, proposal: T::Hash, #[compact] index: ProposalIndex, proposal_weight_bound: Weight
+		) -> DispatchResultWithPostInfo {
 			let _ = ensure_signed(origin)?;
 
 			let voting = Self::voting(&proposal).ok_or(Error::<T, I>::ProposalMissing)?;
 			ensure!(voting.index == index, Error::<T, I>::WrongIndex);
-			ensure!(system::Module::<T>::block_number() >= voting.end, Error::<T, I>::TooEarly);
-
-			// default to true only if there's a prime and they voted in favour.
-			let default = Self::prime().map_or(
-				false,
-				|who| voting.ayes.iter().any(|a| a == &who),
-			);
 
 			let mut no_votes = voting.nays.len() as MemberCount;
 			let mut yes_votes = voting.ayes.len() as MemberCount;
 			let seats = Self::members().len() as MemberCount;
+			let approved = yes_votes >= voting.threshold;
+			let disapproved = seats.saturating_sub(no_votes) < voting.threshold;
+			// Allow (dis-)approving the proposal as soon as there are enough votes.
+			if approved || disapproved {
+				let p = ProposalOf::<T, I>::get(proposal);
+				let proposal_weight = p.as_ref().map(|p| p.get_dispatch_info().weight).unwrap_or(0);
+				ensure!(proposal_weight <= proposal_weight_bound, Error::<T, I>::WrongProposalBound);
+				return Ok(Some(
+					weight_for::close_without_finalize(T::DbWeight::get(), seats)
+						.saturating_add(Self::finalize_proposal(approved, seats, voting, proposal, p))
+				).into());
+			}
+
+			// Only allow actual closing of the proposal after the voting period has ended.
+			ensure!(system::Module::<T>::block_number() >= voting.end, Error::<T, I>::TooEarly);
+		
+			// default to true only if there's a prime and they voted in favour.
+			let default = Self::prime().map_or(false, |who| voting.ayes.iter().any(|a| a == &who));
+		
 			let abstentions = seats - (yes_votes + no_votes);
 			match default {
 				true => yes_votes += abstentions,
 				false => no_votes += abstentions,
 			}
+			let approved = yes_votes >= voting.threshold;
 
+			let p = ProposalOf::<T, I>::get(proposal);
+			let proposal_weight = p.as_ref().map(|p| p.get_dispatch_info().weight).unwrap_or(0);
+			ensure!(proposal_weight <= proposal_weight_bound, Error::<T, I>::WrongProposalBound);
 			Self::deposit_event(RawEvent::Closed(proposal, yes_votes, no_votes));
-			Self::finalize_proposal(yes_votes >= voting.threshold, seats, voting, proposal);
+			Ok(Some(
+				weight_for::close_without_finalize(T::DbWeight::get(), seats)
+					.saturating_add(T::DbWeight::get().reads(1)) // read `Prime`
+					.saturating_add(Self::finalize_proposal(approved, seats, voting, proposal, p))
+			).into())
```

Comment:
I dont like that you have hidden `finalize_proposal` in the weight function.

Reply:
Ok, good point, should be more obvious that logic happening there.
I'll pull it out.